### ISSUE TITLE
PropertyStatus partially not updated after BO Save()

### ISSUE THREAD

**Issue by ck-g** [MarimerLLC/cslaforum#228]
### WPF with csla 4.6.300

Hello,
I have a BusinessRule with a PrimaryProperty and one InputProperty.
When the BO is fetched (DataPortal_Fetch) for display everything works fine:
- WPF PropertyStatus is bind to the PrimaryProperty and is maybe displayed as Warning according to the Rule
- it is updated whenever the PrimaryProperty is changed
- and it is updated whenever the InputProperty is changed
OK

The problem occurs when the BO has been saved (after editing) via Save/BeginSave/ DataPortal_Update:
- now PropertyStatus is not updated whenever the InputProperty is changed
- only after changing the PrimaryProperty everything works (again) as expected.

During debugging I can clearly see that the Rule is executed whenever the InputProperty is changed, nevertheless, PropertyStatus is not being updated (e.g. still displays the Warning message).

Any ideas? 
Did I miss something obvious?

### My Business Rule 
public class AnyRequired : BusinessRule
    {
        private readonly PropertyInfo<int?> _ratingProperty;
        public AnyRequired(IPropertyInfo primaryProperty, PropertyInfo<int?> ratingProperty)
            : base(primaryProperty)
        {
            _ratingProperty = ratingProperty;
            InputProperties = new List<IPropertyInfo> { primaryProperty };
            InputProperties.Add(ratingProperty);
        }
        protected override void Execute(RuleContext context)
        {
            string findings = (string)context.InputPropertyValues[PrimaryProperty];
            int? rating = (int?)context.GetInputValue(_ratingProperty);

            if ((rating == null) & (String.IsNullOrEmpty(findings)))
                context.AddWarningResult("a comment is neccessary!");
        }
    }

**Comment by rockfordlhotka** [comment]
Are you saying that there's a broken rule (warning) when you save, and that the rule _appears_ to still be broken after the save, but during the save you cleared the broken rule list?

**Comment by ck-g** [comment]
Not exactly.
It's a broken rule (warning) when I save, and after the save it is still there (everything fine).
When I change the InputProperty after the save the warning should disappear; but the rule is still broken (the broken rule list is not updated as it seems).
The warning only disappears when I change the PrimaryProperty; changing the InputProperty (after a save) does not alter anything.

**Comment by jonnybee** [comment]
Can you check that the:

- PropertChangedMode is NOT Windows 
- after the InputProperty has changed - is there a BrokenRule With Warning on the PrimaryProperty in BrokenRules? 
- and that after Save you get a PropertyHasChanged on the PrimaryProperty - also when the InputProperty is changed 

It is the PropertyHasChanged event that trigger refresh of PropertyStatus in the UI.

**Comment by ck-g** [comment]
PropertyChangedMode is "xaml". 
I will check the others in a few hours.

**Comment by rockfordlhotka** [comment]
Are you sure you are rebinding the UI to the result of the Save operation? 

Also I guess I don't understand about input/primary property - in your example code you set the input property to be the primary property so they are the same?

**Comment by ck-g** [comment]
@jonnybee 
My findings:
- PropertyChangedMode is "xaml". 
- After the Save I also get a PropertyHasChanged when changing InputProperty or PrimaryProperty
- BrokenRules contains correct rule violations (even after the BO Save), everything fine 
- But the PropertyStatus is not updated; it seems to be out of sync with BrokenRules (unless I again change the PrimaryProperty)

**Comment by ck-g** [comment]
@rockfordlhotka 
My intention was: there is PrimaryProperty which should display a warning if is empty and a dependent property is at the same time null 

Due to your comment I tried it with AffectedProperties (in a private Rule class) instead of DependentProperties:

BusinessRules.AddRule(new FindingsNotEmptyIfRatingNull
                            { PrimaryProperty = P21_FindingsProperty,
                              AffectedProperties = { P21_GenerallyProperty }
                            });

private class FindingsNotEmptyIfRatingNull : Csla.Rules.BusinessRule
        {
            protected override void Execute(Csla.Rules.RuleContext context)
            {
                var target = (ProductEdit)context.Target;

                var findings = target.ReadProperty(P21_FindingsProperty);
                var rating = target.ReadProperty(P21_GenerallyProperty);
                if ((rating == null) & (string.IsNullOrEmpty(findings)))
                    context.AddWarningResult("comment necessary");
            }
        }
		
With this implementation I still have the same problem, PropertyStatus is not updated. Although: in difference to the first code sample this time the BusinessRule is NOT executed when changing the Affected Property after the BO Save. 

I think it has something to do with not properly rebinding the UI to the result of the Save operation, as you suggested.

**Comment by rockfordlhotka** [comment]
It may also be related to the way the PropertyChanged event is raised by the business object.

PropertyChanged is raised for the property that is changed by the user of course. But if your rule affects other properties it is up to your rule to tell CSLA that those other properties are affected so PropertyChanged events are raised for them as well.

So if you have a rule that deals with multiple properties, it will trigger when the rule's PrimaryProperty is changed - and of course that'll cause the primary property's PropertyChanged to be raised because that is the property that changed.

But other input properties used by the rule won't result in PropertyChanged events for those properties because they aren't _affected_ - unless you also add them to the list of properties affected by the rule (the `AffectedProperties` list).

In short:

* PrimaryProperty is the property that triggers the rule
* InputProperties are other properties used by the rule (but not _affected_)
* AffectedProperties are affected by the rule, so PropertyChanged is raised for them

**Comment by rockfordlhotka** [comment]
I tried to replicate the issue but can't - or I am misunderstanding the issue.

Here is my test app: https://1drv.ms/u/s!AjmEoKAAdzm5yaNmOlvgq3Z3b1OClg

**Comment by ck-g** [comment]
Thanks for your feeback and clearing my mind about Primary/Input/AffectedProperties. 
AffectedProperties is definitely the way to go in this case.

I had a look at your test app - and must admit that I can't replicate the issue either.
I think I will try to setup a new sln (copy + paste). Maybe it has something to with my dev env.

**Comment by ck-g** [comment]
For test purposes I added in the ProjectTracker sample a PropertyStatus control each to "Started" and "Ended" in the ProjectEdit view. 
It's not exact the same issue I encounter but when playing around with the dates you can finally observe that although there are no Broken Rules and the BO is allowed to be be saved, the PropertyStatus of "Started" still displays an error.  The error disappears after saving the BO.
I added a screenshot. 
![propertystatus2](https://cloud.githubusercontent.com/assets/15851318/17792311/55df7832-65a0-11e6-97da-429c1a22b6f3.png)

**Comment by ck-g** [comment]
I added view model classes to the test app of rockfordlhotka - now the issue can be replicated. 
It probably has something to do with OnModelChanged(old, new) in the vm. Maybe there is somebody who can explain it. Thanks.

https://1drv.ms/u/s!AjwzsnOajyUMgiSI02mONTK1pleV

The issue in short:
- the test app has one business rule: if both AddressLines in Person are empty, two warnings are displayed for each field via PropertyStatus
- if at least one AddressLine contains an entry NO warning is displayed
- if you click Save, then this rule is somehow broken: if one AddressLine contains an entry the other field still displays the warning (in case two warnings were displayed before clicking Save)

**Comment by rockfordlhotka** [comment]
From this new information it sounds like your problem may be related to this issue: https://github.com/MarimerLLC/csla/issues/600
